### PR TITLE
Fixes and Enhancements for Backend Controllers and Models

### PR DESCRIPTION
**Pull request by chnthkksn**
In this PR, I have made the following modifications:

- Fixed a typo in the budgetController and budgetModel 
- Added the "type" attribute to the categoryModel
- Added "balance", "income", and "expense" attributes to the budgetModel
- Implemented an automatic calculation function in the budgetModel that updates the balance, income, and expense when a new transaction is created, updated or deleted.

### CODE REVIEW

File: server/models/categoryModel.js [hunk 1]
```diff
@@ -21,6 +21,10 @@ Category.init(
       type: DataTypes.STRING,
       allowNull: false,
     },
+    type: {
+      type: DataTypes.ENUM('income', 'expense'),
+      allowNull: false,
```
Comment (nimi-kaushal) [inline server/models/categoryModel.js]:
```
module.exports = {
 
  CategoryType: Object.freeze({
    INCOME: 'I',
    EXPENSE: 'E',
  }),
};

```

and import

File: server/models/transactionModel.js [hunk 2]
```diff
@@ -46,4 +47,50 @@ Transaction.init(
   }
 );
 
+// this hook will be called before creating a new transaction
+Transaction.addHook('afterCreate', async (transaction) => {
+  const budget = await Budget.findByPk(transaction.budgetId);
+
+  if (transaction.type === 'income') {
```
Comment (nimi-kaushal) [inline server/models/transactionModel.js]: use enum CategoryType

**Review by nimi-kaushal — CHANGES_REQUESTED** [review]

File: server/models/transactionModel.js [hunk 3]
```diff
@@ -46,4 +47,50 @@ Transaction.init(
   }
 );
 
+// this hook will be called before creating a new transaction
+Transaction.addHook('afterCreate', async (transaction) => {
```
Comment (nimi-kaushal) [inline server/models/transactionModel.js]: tansfer the logic to tascationService and call here

File: server/models/transactionModel.js [hunk 4]
```diff
@@ -46,4 +47,50 @@ Transaction.init(
...
+  const budget = await Budget.findByPk(transaction.budgetId);
+
+  if (transaction.type === 'income') {
+    budget.income = budget.income + transaction.amount;
+  } else if (transaction.type === 'expense') {
+    budget.expense = budget.expense + transaction.amount;
+  }
+
+  budget.balance = budget.income - budget.expense;
+  await budget.save();
```
Comment (nimi-kaushal) [inline server/models/transactionModel.js]: put` await budget.save();`  to repostory file and call here

File: server/models/transactionModel.js (hunk 2 above)
Comment (chnthkksn) [reply server/models/transactionModel.js]: added ENUMs in commit b65f80fff5815e341e0b7113865a636599e181a1

File: server/models/categoryModel.js (hunk 1 above)
Comment (chnthkksn) [reply server/models/categoryModel.js]: Changed in commit b65f80fff5815e341e0b7113865a636599e181a1

**Comment by chnthkksn** [comment]
- Also I have separated data access logic from controllers to repositories in commit 474a9301c5808b8048d33c75d0c02842a376d3e0
- Separated transaction model hooks from transaction model to service in commit c08c45ff652b45f08979be5d68716c8a23f0e09b
- Refactored code

**Review by nimi-kaushal — APPROVED** [review]

**Pull request merged**